File: components/path-publisher/docker/path_publisher/src/path_publisher_node.cpp


#include <path_publisher/PathPublisher.h>

#include <ros/ros.h>

using namespace std;



bool readWaypoints(vector<geometry_msgs::PoseStamped>& waypoints) {
	
	ros::NodeHandle nodePrivate("~");

	vector<string> waypointsList;


	if (nodePrivate.getParam("waypoints", waypointsList)) {

		geometry_msgs::PoseStamped pose;

		int line = 1;

		for(auto waypointString : waypointsList) {
			double heading = 0;

			auto parsedValues = sscanf(waypointString.c_str(), "%lf,%lf,%lf",
					&pose.pose.position.x,
					&pose.pose.position.y,
					&heading);

			pose.header.frame_id = "map";
			pose.header.stamp = ros::Time(0);

			pose.pose.orientation = tf::createQuaternionMsgFromYaw(heading);

			waypoints.push_back(pose);

			if (parsedValues < 3) {
				ROS_ERROR("Failed to parse a waypoint (line %i)", line);
				return false;
			}

			line++;
		}

	} else {


		ROS_ERROR("Error: waypoints parameter does not exists or empty");
		
		return false;
	}

	return true;
}


int main(int argc, char **argv)
{
	ros::init(argc, argv, "path_publisher_node");

	vector<geometry_msgs::PoseStamped> waypoints;

	if (readWaypoints(waypoints) == false) {
		
		ROS_WARN("Waypoints list is empty, exiting");
		cerr<<" Waypoints list is empty, exiting "<<endl;
		
		return -1;

	} else {


		PathPublisher PathPublisher(waypoints);

		ros::spin();
	}




	return 0;
}

File: components/path-waypoints-navigation/docker/waypoints_navigation_ws/src/path_publisher/src/path_publisher_node.py
#!/usr/bin/env python3
import rospy
import sys
from geometry_msgs.msg import PoseStamped
from nav_msgs.msg import Path
from std_msgs.msg import String
from rospy import Time
import geometry_msgs.msg
from geometry_msgs.msg import PoseStamped
from std_msgs.msg import Bool



class path_publisher:

  def __init__(self):
    self.waypointsList = []
    self.sub = rospy.Subscriber("/move_base_simple/goal",PoseStamped,self.poseCallback, queue_size = 1)
    self.f = open('path.txt','w+')

  
  def poseCallback(self,pose_msg):
      poseStamped = PoseStamped()
      poseStamped.pose.position.x = pose_msg.pose.position.x
      poseStamped.pose.position.y = pose_msg.pose.position.y
      poseStamped.pose.position.z = pose_msg.pose.position.z
      poseStamped.header.frame_id = "map"
      poseStamped.header.stamp = rospy.Time(0)
      poseStamped.pose.orientation = pose_msg.pose.orientation
      orientation_list = [poseStamped.pose.orientation.x,poseStamped.pose.orientation.y,poseStamped.pose.orientation.z,poseStamped.pose.orientation.w]
      

      line = str(poseStamped.pose.position.x) + "," +  str(poseStamped.pose.position.y) + "," +  str(poseStamped.pose.position.z) + "," +  str(poseStamped.pose.orientation.x) + "," + str(poseStamped.pose.orientation.y) + "," + str(poseStamped.pose.orientation.z) + "," + str(poseStamped.pose.orientation.w)
      self.f.write(line + "\n")

    
def publishPath():
    path_pub = rospy.Publisher('/waypoints_route', Path, queue_size=10)
    route = Path()
    route.header.frame_id = "map"
    route.header.stamp = rospy.Time(0)
    f = open("path.txt")
    for line in f:
        print(line.rstrip())
        pose_msg = line.split(",")
        pose = PoseStamped()
        pose.header.frame_id = "map"
        pose.pose.position.x = float(pose_msg[0])
        pose.pose.position.y = float(pose_msg[1])
        pose.pose.position.z = float(pose_msg[2])
        pose.pose.orientation.x = float(pose_msg[3])
        pose.pose.orientation.y = float(pose_msg[4])
        pose.pose.orientation.z = float(pose_msg[5])
        pose.pose.orientation.w = float(pose_msg[6])
        route.poses.append(pose)
    while not rospy.is_shutdown():
        path_pub.publish(route)
    print("out")    
        
    


if __name__ == '__main__':
    rospy.init_node('path_publisher_node', anonymous=True) 
    
    """if the user didn't enter and argument the node will assume that the 'path.txt' file exists 
    and will publish path message according to the points that are written in the file"""
    if len(sys.argv) != 2:
        publishPath()
        rospy.spin()
    else:
        """if the user entered the number 1 as an argument to the node, the node will read from the file ('path.txt') the points
        and will publish a path message according to those points. If the argument to the node was 0, the node will create path and write the points in the 'path.txt' file """
        if int(sys.argv[1]) == 1:
            print("Publish Path")
            publishPath()
            rospy.spin()
        elif int(sys.argv[1]) == 0:
            print("Create Path")
            pathPublisher = path_publisher()
            rospy.spin()
            pathPublisher.f.close()
            print("Closed File")

